src/beamlime/applications/groups.py

```diff
@@ -0,0 +1,225 @@
+# SPDX-License-Identifier: BSD-3-Clause
+# Copyright (c) 2023 Scipp contributors (https://github.com/scipp)
+
+import asyncio
+from logging import Logger
+from typing import Callable, Optional, TypeVar
+
+from .. import protocols
+from ..core.schedulers import async_timeout
+from .interfaces import BeamlimeApplicationInterface
+from .mixins import LogMixin
+
+_CommunicationChannel = TypeVar(
+    "_CommunicationChannel",
+    protocols.BeamlimeDownstreamProtocol,
+    protocols.BeamlimeUpstreamProtocol,
+    protocols.BeamlimeTwoWayProtocol,
+)
+
+MAX_INSTANCE_NUMBER = 2
+
+
+class ApplicationStoppedException(Exception):
+    ...
+
+
+class ApplicationNotStartedException(Exception):
+    ...
+
+
+# TODO: Implement multi-process/multi-machine instance group interface.
+class BeamlimeApplicationInstanceGroup(LogMixin):
+    """
+    Multiple instances of a single type of application.
+    It overwrites all interfaces of ``BeamlimeApplicationInterface``,
+    in a way that it has the exact same interfaces
+    just like a single application but controls multiple instances internally.
+    Therefore it does not inherit ``BeamlimeApplicationInterface``,
+    but it should pass the protocol test.
+    See ``protocol_test.test_instace_group_protocol`` of the unit test package.
+
+    Parameters
+    ----------
+    constructor: Callable
+        ``constructor`` should be able to construct an instance
+        by calling it without any arguments.
+        It can be made by ``functools.partial``.
+        See the example usage in the ``BeamlimeSystem.build_instances``.
+    instance_num: int
+        Number of instance the ``ApplicationInstanceGroup`` will populate and carry.
+        It initially populates instances of ``instance_num``.
+    app_name: str
+        Name of the application.
+
+    Protocol
+    --------
+    BeamlimeApplicationProtocol
+
+    """
+
+    def __init__(
+        self,
+        constructor: Callable,
+        instance_num: int = 2,
+        logger: Logger = None,
+        app_name: str = None,
+    ) -> None:
+        self.app_name = app_name or ""
+        # TODO: Add method to check ``_instances`` if all instances are alive.
+        self._init_logger(logger=logger)
+        self._init_instances(constructor, instance_num)
+        self._input_ch = None
+        self._output_ch = None
+        self._tasks = []
+        self._timeout = 10
+        self._update_rate = 1
+
+    def _init_logger(self, logger: Logger) -> None:
+        if logger is None:
+            from ..logging import get_logger
+
+            self.logger = get_logger()
+        else:
+            self.logger = logger
+
+    def _init_instances(self, constructor: Callable, instance_num: int) -> None:
+        self.constructor = constructor
+        self._instances = []
```

Comment:
I think maybe a `dict` of instances instead of a `list` may allow for easier debugging if you want to inspect a particular instance?